confmap/resolver.go

```diff
@@ -27,18 +27,35 @@ type Resolver struct {
 	watcher chan error
 }
 
+type ConverterFactory = func(ConverterSettings) Converter
+
+type ProviderFactory = func(ProviderSettings) Provider
```

Comment:
I assume we will probably want to make this closer to `exporter.Factory` and the like, but I am fine doing that in a different PR

Reply:
That's a good idea, making this an interface would make these modules more flexible going forward. I'm okay if we want to tackle that first, but I'm also okay doing it later since it will let us more effectively use `ConverterSettings` and `ProviderSettings` in the meantime.

Reply:
Thinking about this a bit more, while I am not too worried about breaking these APIs, I think we should tackle it first to avoid further complexity on the builder (we would have to check `otelcol_version` and generate different code depending on the version chosen)

Reply:
I missed this comment, I'll make this update shortly.

Reply:
Okay, I've updated all of the confmap modules to be created using factories. It's a little heavy, but I think it works. Let me know what you think.